hack/lib/golang.sh

```diff
@@ -441,6 +450,98 @@ kube::golang::outfile_for_binary() {
   echo "${output_path}/${bin}"
 }
 
+# Argument: the name of a Kubernetes package.
+# Returns 0 if the binary can be built with coverage, 1 otherwise.
+# NB: this ignores whether coverage is globally enabled or not.
+kube::golang::is_covered_binary() {
+  return $(kube::util::array_contains "$1" "${KUBE_COVERAGE_INSTRUMENTED_PACKAGES[@]}")
+}
+
+# Argument: the name of a Kubernetes package (e.g. k8s.io/kubernetes/cmd/kube-scheduler)
+# Echos the path to a dummy test used for coverage information.
+kube::golang::path_for_coverage_dummy_test() {
+  local package="$1"
+  local path="${KUBE_GOPATH}/src/${package}"
+  local name=$(basename "${package}")
+  echo "$path/zz_autogenerated_${name}_test.go"
+}
+
+# Argument: the name of a Kubernetes package (e.g. k8s.io/kubernetes/cmd/kube-scheduler).
+# Creates a dummy unit test on disk in the source directory for the given package.
+# This unit test will invoke the package's standard entry point when run.
+kube::golang::create_coverage_dummy_test() {
+  local package="$1"
+  local name="$(basename "${package}")"
+  cat <<EOF > $(kube::golang::path_for_coverage_dummy_test "${package}")
+package main
+import (
+  "testing"
+  "k8s.io/kubernetes/pkg/util/coverage"
+)
+
+func TestMain(m *testing.M) {
+  // Get coverage running
+  coverage.InitCoverage("${name}")
+
+  // Go!
+  main()
+
+  // Make sure we actually write the profiling information to disk, if we make it here.
+  // On long-running services, or anything that calls os.Exit(), this is insufficient,
+  // so be sure to call this from inside the binary too.
+  coverage.FlushCoverage()
+}
+EOF
+}
+
+# Argument: the name of a Kubernetes package (e.g. k8s.io/kubernetes/cmd/kube-scheduler).
+# Deletes a test generated by kube::golang::create_coverage_dummy_test.
+# It is not an error to call this for a nonexistent test.
+kube::golang::delete_coverage_dummy_test() {
+  local package="$1"
+  rm -f $(kube::golang::path_for_coverage_dummy_test "${package}")
+}
+
+# Arguments: a list of kubernetes packages to build.
+# Expected variables: build_args should be set to an array of Go build arguments.
+# In addition, the standard build globals are assumed.
+#
+# Invokes Go to actually build some packages. If coverage is disabled, simply invokes
+# go install. If coverage is enabled, builds covered binaries using go test, temporarily
+# producing the required unit test files and then cleaning up after itself.
+# Non-covered binaries are then built using go install as usual.
+kube::golang::build_some_binaries() {
+  if [[ -n "${build_with_coverage:-}" ]]; then
```

Comment:
This and the phrase "standard build globals" sent me down a twisty rabbit hole I wish I hadn't fell into.

It seems like it'd be simpler to use `KUBE_BUILD_WITH_COVERAGE` here directly instead of capturing it as a local, uh, global below?

Reply:
I did this in an attempt at consistency with the all the other variables that get used here. I don't really have strong opinions, and I agree that it'd be slightly simpler to not capture `build_with_coverage`, but I think it would also be inconsistent?

Reply:
Yeah it looked like following the consistency for consistency's sake, I just don't know if we need to keep cargo culting it.  The reason for the capture looked to me like conveniently munging lists of flags.  This isn't that, and it's also not a well known env var used by other tools, but one we made up.  So in this case I would follow the pattern of doc'ing this as a parameter to this function.

That said, I would like to see if we can get this included in the release, so I'll file an issue for this nit.